### ISSUE TITLE
Plan for breaking changes in Shelley `ApplyTxErr` CBOR

### ISSUE THREAD

**Issue by nfrisby** [input-output-hk/ouroboros-network#2488]
PR input-output-hk/cardano-ledger-specs#1740 would enable the straight-forward and desired design in the Cardano ThreadNet tests. But the PR makes changes to CBOR instances that `ouroboros-shelley` reuses, and thereby changes our binary compatibility.

That PR makes two specific changes.

- A constructor of `PredicateFailure` -- eventually included in `ApplyTxError` -- gains an argument. (We could instead add a new partially-redundant constructor, if that would be easier to handle.)
- The verkey and signature in a bootstrap witnesses is now fixed to Ed25519: size of 32 and 64 bytes respectively. This is not a change compared to "real" Shelley crypto, but is for mock crypto used in tests, etc.

This Issue tracks determining how to mitigate the corresponding breaks of binary compatibility for downstream users of `ouroboros-shelley` and `ouroboros-cardano`. It'd be particularly nice if the solution could avoid similar fragility in the future.

Discussion from @mrBliss on Slack `#ouroboros-prototyping`: https://input-output-rnd.slack.com/archives/GCW2S19CM/p1596175870006300 https://input-output-rnd.slack.com/archives/GCW2S19CM/p1596176711012900

**Comment by nfrisby** [comment]
This blocks PR https://github.com/input-output-hk/cardano-ledger-specs/pull/1740, which in turn blocks Issue #2388.

**Comment by mrBliss** [comment]
@nfrisby Is this still needed now that you have a workaround in #2502?

**Comment by nfrisby** [comment]
Yep. This still blocks the path that would be more preferable to the workaround.

Status: There was mention of a spreadsheet that I could include this breaking change for the next "batch" of breaking changes. I'm still working to get this particular change as an item on that spreadsheet -- I requested spreadsheet access, but haven't followed-up with a repeat request.

**Comment by nfrisby** [comment]
Thomas briefly summarized his related plans in a `#rule-based-spec-sl` thread: https://input-output-rnd.slack.com/archives/CCRB7BU8Y/p1599838139132600?thread_ts=1599794244.131000&cid=CCRB7BU8Y

**Comment by mrBliss** [comment]
I have labelled this issue with "transition2" so that we don't forget it when the next hard fork happens.

**Comment by mrBliss** [comment]
@dcoutts: I had a realisation last night: we only serialise `ApplyTxErr` to send it across the `LocalTxSubmission` protocol. This means that only the wallet or db-sync (are there other clients I'm missing?) will try to deserialise it. *And isn't it always the case that the local node the wallet/db-sync talks to is installed/deployed together with the wallet/db-sync, i.e., using the same version of `cardano-ledger-specs`?* If that is true, then we are free to change the `ApplyTxErr` CBOR encoding without having to worry about compatibility, right?

In fact, we wouldn't even have to worry about versioning the `NodeToClient` :thinking:. Or am I being daft and forgetting something obvious? (Or can't we assume they are using the same version of `cardano-ledger-specs`?)
